docker: Build with Alpine Linux

andy5995 opened the pull request:
For #50

[skip ci]

This image will build, but swift is not available

[SR-4632] Swift on Alpine Linux (support musl libc) https://github.com/swiftlang/swift/issues/47209

Image sizes:

This is 1.1GB, the Debian image is currently 3.83

This PR can be closed and use for future use.

andy5995 committed:
docker: Build with Alpine Linux

For #50

[skip ci]

This image will build, but swift is not available

[SR-4632] Swift on Alpine Linux (support musl libc)
https://github.com/swiftlang/swift/issues/47209

Image sizes:

This is 1.1GB, the Debian image is currently 3.83

This PR can be closed and use for future use.

KaruroChori commented:
That would be very bad. Supporting musl libc is not really negotiable, but my understanding was that Swift 6 was capable.  
I need to investigate this more in depth.

andy5995 commented on docker/Dockerfile:
```diff
@@ -1,35 +1,35 @@
+# https://github.com/swiftlang/swift/issues/47209
+# Swift on Alpine not yet supported
+#
+# RUN curl -s https://archive.swiftlang.xyz/install.sh | bash && \
```
When this ran, I got the error that the operating system was unsupported.

KaruroChori replied:
musl support is an open issue under certain conditions, but that looks like more of an arbitrary restriction of the installer script. I need to test it in a real env.

KaruroChori replied:
I get it. Swift can crosscompile to support musl and so alpine as target. It cannot run on it yet.

andy5995 replied:
Ok. For reference: [Building from source](https://swift-linux.readthedocs.io/en/latest/others.html#building-from-source) or the options above it. That could be done in the Dockerfile.

KaruroChori commented:
Oh no you nuked it! :D. I was keeping it for reference since I closed the issue.

KaruroChori commented:
In https://github.com/KaruroChori/vs-fltk/pull/88 I removed swift. 
vsfltk will continue to support it for linking, but it is not going to be used in this repository (but it is very likely it will be used for an external repo of components).
Hence, it is possible to rebase the docker image on alpine if so desired.

andy5995 committed:
docker-compose.yml: Add build option

andy5995 commented:
Current state after I've just rebased and updated it a little:

## 1

> scripts/build-default.sh: line 48: bun: not found

The path is in ~/.profile but it's not getting sourced automatically. The command in entry-default.sh might need to be tweaked for Alpine

`su builder` Adding -l or some-such nonsense but I haven't investigated yet ;)

## 2

```
Build type: native build
Project name: vs-fltk
Project version: 0.1.2-alpha

meson.build:1:0: ERROR: Unsupported linker, only bfd, gold, and lld are supported, not lld-20.
```

Which is obviously due to the version (21) set in platforms/clang.ini

## 3

When I try to execute sudo:

bash: sudo: command not found

Anyway, I'll wait to hear comments about this from you before working on it further.

KaruroChori commented:
Hi, thanks for pushing this forward! I am trying to figure out a way to more gracefully handle the `native` and `crossbuild` files around my projects, so to avoid issues related to the compiler version and duplicating over and over.  
The lack of `sudo` is also expected, alpine is quite minimalistic when distributed as a base image. It would need to be installed as an extra package.

KaruroChori commented:
https://github.com/KaruroChori/vs-fltk/actions/runs/16065997386/job/45340545225 the docker workflows are back working as intended, so whatever is left not working should be specific for alpine.

andy5995 committed:
Dockerfile: Fix 'bun: not found'

andy5995 commented:
> The lack of `sudo` is also expected, alpine is quite minimalistic when distributed as a base image. It would need to be installed as an extra package.

That's my problem. In the Dockerfile, it's in the list of files that get installed. ;) I'm investigating now...

andy5995 committed:
Dockerfile: Fix sudo not found (fix comment location)

andy5995 committed:
Dockerfile: Install lld and make

andy5995 committed:
entry-build.sh: source ~/.bashrc

andy5995 commented:
The docker stuff is working now, but the [libriscv build fails](https://github.com/KaruroChori/vs-fltk/actions/runs/16260943326/job/45905853137#step:8:1266).

KaruroChori commented:
That is strange, it seems like it cannot find a syscall mapped in musl. I will try to update the dependency and if it does not work I am going to notify the issue upstream.

andy5995 commented:
> That is strange, it seems like it cannot find a syscall mapped in musl. I will try to update the dependency and if it does not work I am going to notify the issue upstream.

I'm testing on HEAD, which is what's specified in the wrap file. I opened a ticket at https://github.com/libriscv/libriscv/issues/301

andy5995 commented:
@KaruroChori , @fwsGonzo fixed the issues and the Alpine build is working now.
